xapian-letor/api/feature_internal.h

```diff
@@ -0,0 +1,151 @@
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301
+ * USA
+ */
+
+#ifndef XAPIAN_INCLUDED_FEATURE_INTERNAL_H
+#define XAPIAN_INCLUDED_FEATURE_INTERNAL_H
+
+#include "xapian-letor/feature.h"
+#include "xapian-letor/featurelist.h"
+
+#include <map>
+
+namespace Xapian {
+
+/** Class defining internals of Feature class. */
+class Feature::Internal : public Xapian::Internal::intrusive_base {
+    friend class Feature;
+
+    /// Xapian::Database using which features will be calculated.
+    Database feature_db;
+
+    /// Xapian::Query using which features will be calculated.
+    Query feature_query;
+
+    /// Xapian::Document using which features will be calculated.
+    Document feature_doc;
+
+    /// Frequency of the Query Terms in the specified documents.
+    std::map<std::string, Xapian::termcount> termfreq;
+
+    /// Inverse Document Frequency of Query terms in the database.
+    std::map<std::string, double> inverse_doc_freq;
+
+    /** Length of the document as number of terms for different parts like
+     *  'title', 'body' and 'whole'.
+     */
+    std::map<std::string, Xapian::termcount> doc_length;
+
+    /** Length of the collection in number of terms for different parts like
+     * 'title', 'body' and 'whole'.
+     */
+    std::map<std::string, Xapian::termcount> collection_length;
+
+    /// Frequency of the Query Terms in the whole database
+    std::map<std::string, Xapian::termcount> collection_termfreq;
+
+  public:
+    /// Default constructor
+    Internal() {}
+
+    /// Constructor creating an object instantiated with db,query and doc
+    Internal(const Xapian::Database& db, const Xapian::Query& query,
+	     const Xapian::Document& doc)
+	 : feature_db(db), feature_query(query), feature_doc(doc) {}
+
+    /// get database
+    Database get_database() const {
+	return feature_db;
+    }
+
+    /// get query
+    Query get_query() const {
+	return feature_query;
+    }
+
+    /// get document
+    Document get_document() const {
+	return feature_doc;
+    }
+
+    /// Get termfreq
+    Xapian::termcount get_termfreq(const std::string& term) const;
+
+    /// Get inverse_doc_freq
+    double get_inverse_doc_freq(const std::string& term) const;
+
+    /// Get doc_length
+    Xapian::termcount get_doc_length(const std::string& term) const;
+
+    /// Get collection_length
+    Xapian::termcount get_collection_length(const std::string& term) const;
+
+    /// Get collection_termfreq
+    Xapian::termcount get_collection_termfreq(const std::string& term) const;
+
+    /** Sets the termfrequency that is going to be used for
+     *  Feature building.
+     *
+     */
+    void set_termfreq(const std::map<std::string,
+			     Xapian::termcount>& tf) {
+	termfreq = tf;
+    }
+
+    /** Sets the inverse_doc_freq that is going to be used for
+     *  Feature building.
+     *
+     */
+    void set_inverse_doc_freq(const std::map<std::string,
+				     double>& idf) {
```

Comment:
Wrapped template arguments ought to align vertically just like wrapped function arguments (so `double` under `std::string`).  There seem to be quite a lot of cases of this in this patch.

Reply:
fixed in a fixup commit.